### ISSUE TITLE
Device Bound Session Credentials

### ISSUE THREAD

**Issue by MrBrain295** [mozilla/standards-positions#912]
## Request for Mozilla Position on an Emerging Web Specification

* Specification title: Device Bound Session Credentials
* Specification or proposal URL (if available): 
* Explainer URL (if available): https://github.com/WICG/dbsc/blob/main/README.md
* Proposal author(s) (`@`-mention GitHub accounts): @kristianmonsen
* Caniuse.com URL (optional): 
* Bugzilla URL (optional): 
* Mozillians who can provide input (optional): 

### Other information

In their words "Device Bound Session Credentials (DBSC) aims to reduce account hijacking caused by cookie theft. It does so by introducing a protocol and browser infrastructure to maintain and prove possession of a cryptographic key. The main challenge with cookies as an authentication mechanism is that they only lend themselves to bearer-token schemes. On desktop operating systems, application isolation is lacking and local malware can generally access anything that the browser itself can, and the browser must be able to access cookies. On the other hand, authentication with a private key allows for the use of system-level protection against key exfiltration."

**Comment by simon-friedberger** [comment]
The proposal seems sound but it's not clear that it is a good investment of engineering time. I've added some comments here: https://github.com/WICG/dbsc/issues/13

**Comment by april** [comment]
I am only a former Mozillian so please take my requests with the proper weighting, but I would love to see Mozilla support DBSC (or the related Microsoft standard.)

Over here at Dropbox we have been seeing an increasing prevalence of malware that steals cookie sessions from cookie stores, and a technology such as DBSC would go a long ways towards mitigating the impact this has on our users.

Thanks, and hope y'all are well.  :)

**Comment by kmonsen** [comment]
> This should also be aligned with Microsoft trying to do something similar in https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/BindingContext/explainer.md

FYI: Microsoft have moved their BPOP effort to be part of DBSC now and we are actively working together: https://github.com/WICG/dbsc/blob/main/DBSCE/Overview.md

**Comment by hpvd** [comment]
+1 on supporting this since it is great step to more security in a place that is harder to control/protect on other ways...

**Comment by ZenithVal** [comment]
Now that Chrome has integrated DBSC, is there hope we'll see this in Firefox sooner rather than later?

**Comment by IOnlyFetchBranches** [comment]
Session hijacks are only becoming *more* common. To be quite honest, i'm not sure why DBSC (or a similar standard) hasn't been more urgently pushed. Would absolutely love if Firefox could support this !

**Comment by Liwindo** [comment]
Worth revisiting this two years on — the landscape has shifted:

1. DBSC has shipped. Chrome 146 (April 2026) rolled it out to stable
   on Windows, macOS to follow. Edge consolidated its BPOP work into
   DBSC. The W3C WebAppSec spec is active and has evolved since 2023
   (cross-origin bindings, federated-identity support, DBSCE variant).

2. The threat model is now empirically validated. Chrome's App-Bound
   Encryption (Chrome 127, July 2024) was bypassed by multiple
   infostealer families within 45 days. Google itself now states
   publicly that "there is no reliable way to prevent cookie
   exfiltration using software alone on any operating system" — which
   is essentially Mozilla's position from bug 56788, two decades on.
   DBSC is the architectural response: move bearer tokens to
   hardware-bound proof-of-possession.

3. That changes the calculus behind @simon-friedberger's 2024
   assessment. Firefox users on Windows and macOS are on track to be
   the only major-browser cohort without a path to hardware-bound
   sessions, on platforms where TPM 2.0 and Secure Enclave are
   baseline. Infostealer-as-a-service has industrialized exactly this
   attack class, and session-token theft is now the dominant MFA
   bypass in incident reports.

Concrete asks:
- A refreshed Mozilla position against the current spec, not the 2023
  explainer.
- If the engineering-cost concern stands, a public articulation of
  what would change it (server-side adoption thresholds? narrower
  MVP? joint work with Microsoft on the converged spec?).
- Tracking bugs in Bugzilla so contributors and security-focused
  downstreams have visibility.

Not asking for a shipping commitment — asking for the position to be
re-evaluated against 2026 evidence rather than 2023 assumptions.

**Comment by ghost** [comment]
how about linux?

this app bound stuff is not implemented in Linux. neither chrome based neither gecko based.

but then again

if malware gets on your PC it is no longer your PC and your data. prevention is better. 

windows app bound is already bypassed. it does basically nothing. 

i do not see a point other then making it harder for malware authors, but then look on GitHub and see plenty of examples. For the average malware author this is not hard at all to bypass.

i consider it as "mopping with the tap running".  we still leave the key under the mat before the front door. 

"Google itself now states
publicly that "there is no reliable way to prevent cookie
exfiltration using software alone on any operating system" — which
is essentially Mozilla's position from bug 56788, two decades on."

this. that is the point of it. it is just useless. as long as it is on your device and not encrypted again with a user master password that must be manually entered, any kind of protection is futile

might as well not encrypt at all and leave all keys in plaintext.  it takes around a few hundred lines of go code to exfill it all.
as i said, if it is on your computer is is no longer your computer and data.

websites should implement systems that detect new devices or strange behavior that detects it. it is not normal that a new browser on a new os in a other country other then the normal user usage is logged in. 

it is the websites who should implement better cookie security. witch would solve a lot of problems. that is not a thing a browser is responsible for.

i would not waste time on this.

**Comment by KyeRussell** [comment]
I do not think that "if you're 'on someone's computer', it's game over" is a threat model that's suitable for the modern landscape. Just recently we saw Microsoft fix Edge unnecessary keeping plain-text passwords in memory.

**Comment by Liwindo** [comment]
Two distinctions worth separating, since the "malware on the device = game over, so why bother" framing keeps recurring:

Read access vs. replayability
A bearer token is valuable precisely because it's portable: exfiltrate once, replay from anywhere, at scale - that's the entire economics of infostealer-as-a-service. A key bound to a TPM / Secure Enclave can still be worked around locally, but it can't be exported and replayed off-device. That doesn't make local compromise harmless; it makes the stolen artifact non-fungible, which is where the actual damage lives. The ABE bypasses are real, but note what they require now, privilege escalation, code injection, or on-host debugger attachment, versus the old model of grabbing a file and selling it in bulk. Raising cost and forcing noisier, on-host, detectable behavior is the point, not perfect prevention.

"Websites should handle this, not the browser."
DBSC is the mechanism by which they do. It's server-opt-in; the browser only provides the hardware-key primitive the server binds its session to. Advocating server-side session hardening while opposing DBSC is opposing the thing that enables it.

The Linux gap is a fair point as there's no equivalent primitive widely deployed there today, and an honest position has to acknowledge that. But that's an argument for scoping, not for declining the position outright.

**Comment by simon-friedberger** [comment]
Proposed position: **negative**

We are generally in favor of providing a way to leverage cryptographic hardware to protect sessions. However, the DBSC design has issues, the most important being:
1. It leaves a window in which an attacker can use a stolen session cookie. Alternative designs which require signatures on every request were not sufficiently explored. Those options could include some form of delegation to shield the TPM from high load if that is necessary.
2. The separate reauthentication flow requires a bespoke protocol which does not match existing cookie handling.

In principle, where there is a TPM available, we don’t mind that being used to back this.  However, we don’t want to see a situation where this is expected.  Especially if that comes with some device attestation that the website receives.  People should be able to choose what software and hardware they use and exposing attestations to the web would undermine that.

A negative position does not necessarily mean that we will not implement a proposal.

**Comment by wparad** [comment]
> It leaves a window in which an attacker can use a stolen session cookie. Alternative designs which require signatures on every request were not sufficiently explored.

💯 Well said.

Negative on this proposal is the right call, because it doesn't solve this critical problem.

**Comment by IOnlyFetchBranches** [comment]
@simon-friedberger I see, and I understand that the current DBSC proposal has shortcomings.

That said, you mentioned that Mozilla is generally in favor of providing hardware-backed protection for sessions. Have you discussed any alternative approaches that would address these concerns while still providing an additional layer of protection?

If there are viable alternatives, I'd be genuinely interested in reading about them. If not, I'm not convinced that a negative position on the current proposal is necessarily the best outcome. Even if DBSC doesn't eliminate every attack, reducing the attack surface still seems like a meaningful improvement over today's bearer-cookie model no?

Regarding the concern about websites coming to expect TPM-backed credentials or device attestations, couldn't that be addressed at the implementation level? For example, Firefox could expose the same DBSC functionality without guaranteeing that keys are hardware-backed or exposing any form of device attestation to websites. That would seem to preserve user choice while still allowing browsers to take advantage of stronger protection where it is available.

I'm curious whether alternatives like that have been considered, or whether there are technical reasons they wouldn't be sufficient. Again, you all are the experts, so I'm genuinely interested in understanding the reasoning behind the decision. Thanks!

**Comment by arnar** [comment]
I do want to call out that the DBSC proposal does not call for any form of attestation, and similarly argues against it. The Chrome implementation provides no attestation or guarantees about the method used to secure the binding key.

Regarding approaches that validate the key on each request: We did explore this at Google, and found this infeasible. We received similar feedback from partners (who I can ask to +1 this if that makes any difference). We expect this will become even harder to do at scale with PQC.

**Comment by fivetenth** [comment]
https://en.wikipedia.org/wiki/Perfect_is_the_enemy_of_good

Very disappointing that my favorite browser is actually the least secure browser for this issue.

**Comment by hpvd** [comment]
just another pov on this:
step by step having no DBSC (or something similar/superior) nor addressing this field actively and positively,
may also becoming an acceptance and markting problem for firefox, because there is a growing number of articles which mark Firefox directly or indirectly as "unsecure" describing the cookie stealing problem in comparison to chrome:

https://www.notebookcheck.net/Account-hijacked-despite-2FA-a-stolen-cookie-is-enough.1364307.0.html
https://www.notebookcheck.com/Konto-weg-trotz-Zwei-Faktor-Der-geklaute-Cookie-reicht.1363771.0.html

https://cyberhoot.com/blog/hackers-steal-your-cookies-chrome-stopped-letting-that-work/

https://www.corbado.com/blog/device-bound-session-credentials-dbsc

**Comment by dluxhu** [comment]
A simple user here and my opinion may not be popular. Google protects the Takeout archives with DBSC (maybe it protects all sessions, I don't know, I never tried it out) on Chrome, so it is impossible to write a script that downloads the archives in a headless computer.

Those are big files and there's a lot of them.

Google failing to provide a reliable and simple way to download Takeout archives requires the users using scripts to avoid clicking on tens of files and manually monitor their download and it's even more complicated if you don't have enough space on your laptop and you want to do it on your NAS.

The curl download links generated from Chrome DevTools are useless (on a Mac) and the only useful browser to work around this is Firefox. Please do not implement DBSC for Firefox (or let us turn it off if we want), I need this browser for my Takeout until Google is willing to give us a better way for headless access to the archives (which will probably never going to happen, because this project seems to be very low priority). Thanks for the attention.

**Comment by cristiancastineiras** [comment]
i understand mozilla's concerns regarding the current design, especially around the replay window and the desire to avoid making tpm-backed credentials or device-specific characteristics an expectation on the web.

however, i think there is a risk of evaluating dbsc against an idealized solution rather than against the current reality.

today, the default session model on the web is still based on bearer tokens. once a session cookie is exfiltrated, it can generally be replayed from anywhere until expiration or revocation. dbsc does not completely eliminate session theft, but it significantly reduces the value of stolen session artifacts by making them non-portable.

https://workspace.google.com/blog/identity-and-security/defending-against-account-takeovers-top-threats-passkeys-and-dbsc

while it is true that an attacker may still have a limited opportunity to use a stolen session before reauthentication occurs, that is still a meaningful improvement over the current model where a cookie can be replayed indefinitely from another device. security improvements are often incremental, and reducing the scale and profitability of attacks matters even when they do not completely eliminate them.

regarding per-request signatures, i agree that such an approach would offer stronger guarantees. however, if large-scale deployability is the main challenge, a practical solution that can be deployed today may provide more real-world benefit than a theoretically stronger design that web services are unlikely to adopt.

i also want to point out that the current proposal explicitly avoids device attestation and does not provide websites with guarantees about the hardware protecting a key. that seems consistent with mozilla's long-standing position on user choice and anti-fingerprinting.

would consider a neutral position if the specification evolved further, or are there fundamental architectural concerns that would remain even if the replay window and protocol complexity were addressed?

**Comment by Lightningbulb2** [comment]
@simon-friedberger 
Not to intrude on the conversation, but I am a random outsider that only ended up here because I was dumb, and my session token got stolen.

I use Firefox as my main browser, and despite windows immediately flagging and removing the malware. My Gmail session token was already stolen and used AFTER my computer was off for 3 hours. They changed the email and password on my EA account, Rockstar, Microsoft, Discord, Epic games, and alternate steam account.

I have 2FA enabled on my Google account, Github, and main Steam account. So those were untouched. But already being logged into Google gave access to SO much, and they could have done way more damage given some more time. Heck, they might have harvested thousands of emails and gotten so much information on my identity and records.

Looking into it, I come to find out that this could have all been avoided if I was using Chrome which is a little frustrating.

**Comment by wparad** [comment]
@Lightningbulb2, it's worth identifying whether or not this change here would have helped you though. Your situation may or may not have been able to be prevented with DBSC. So the real question is, did the attacker use a stolen session token acquired through an exploit by having to your machine or only to the session cookie?

If someone has access to your machine, they can masquerade as you and no additional security implementation can stop this. But if only the browser data is leaked that is a different story.

Would you be able to share more about how you believe access to your Google account was compromised?

**Comment by Lightningbulb2** [comment]
@wparad 

Of course!

I am confident it is purely a token exploit because I turn my desktop computer OFF at night. So that would be the only way for them to get in. I don't have hibernate or sleep even enabled. So forgetting would mean my room is all lit up and my fans noisily spinning beside my bed.

This is also why I even looked into this in the first place. Because I was surprised they could use my session token from a different computer, I thought voiding stuff like that was already a thing.

***The entire thing happened with my computer offline, and I was woken up by my brother spam calling my phone because I was sending scam screenshots to everyone on discord.***

**Comment by wparad** [comment]
But that's still not enough to narrow it down. DBSC doesn't prevent using a token a when your computer is off. It only prevents using a token that was stolen when your computer is on, but it doesn't prevent the attacker from generating a new one while your computer is on. Do you know how your machine was compromised? That's the only way to know if DBSC would have helped.

**Comment by Lightningbulb2** [comment]
@wparad 
I'm not super in the know, but I don't see how they could have generated a new one. The malware I downloaded was only acquired THAT night and I never logged into anything, so a fresh token was never generated as far as I know.

But if the malware can arbitrarily request a fresh token to use on another device without authentication, that would sound like a much larger security hole, and would also be impossible for me to tell the difference between.

I ran sketchy .EXEs that windows immediately flagged, that's the limit to what I know.

Unless there is some sort of way to see my login/cookie history somewhere.

**Comment by wparad** [comment]
If you had captured a list of sessions from Google, you could check if a new session was created, but as DBSC doesn't prevent an attacker from generating a new session, the best way to know especially since you had MFA on your account is to remember if you ever were forced to reauthenticate since the malware was installed.

Given the relative recency of DBSC, you are probably right that it would have prevented this attack, but given that attackers will just updote the target to forcing a relogin it the future, the actions you took would still lead to future compromise. That's because DBSC doesn't prevent a malware from binding your new login to a device they control.

**Comment by Lightningbulb2** [comment]
@wparad 

You are right in that while it IS a security hole. Undetected malware can still just wait or force you to login.

I agree on that, but in my case during one night. I tried running malware, it was detected, I never logged into anything or re-authenticated, and my gmail access was compromised hours later while my computer was off. It got caught! The anti-virus did its job, but that glance at my computer was able to do so much.

Checking my Google account sessions, I am actually accidentally testing my hypothesis right now.

I had clicked "Sign out of all accounts" on gmail thinking it would sign out on all my devices, but my activity shows I'm still logged in on everything. Which means I have only re-logged into my firefox but not my Chrome session, and I haven't forced out my Linux dual-boot sessions or my phone.

<img width="665" height="395" alt="Image" src="https://github.com/user-attachments/assets/fd19fa8c-62c1-4b7a-99e9-ca9f1ca8470c" />

So IF they did get a login, it would show up, and google chrome is protected from this attack so those cookies are fine if stolen.

The attack was 2 days ago, and they haven't done anything after I only reset my firefox session. (A little accidental risky test for ya, oops)

**Comment by arnar** [comment]
This probably isn't the right place to dissect an individual compromise, but a quick re-statement of Google's motivating factors for DBSC since this case is a good demonstration of those factors: 

This type of drive-by infostealer malware scales very easily, even if it is detected and cleaned up almost immediately the damage is already done. There is almost no opportunity for sign-in processes or on-device antivirus to address it. If we force such malware to instead to recreate new sessions or do its abuse locally on your device, many other opportunities open up for fighting that.

DBSC alone won't fix the malware problem, but we think it is a necessary piece in the overall puzzle.
